### PR TITLE
Percentile Function

### LINKED ISSUE #17469: <Percentile>

### PR DESCRIPTION
**Pull request by dhruvsaxena11**
Hi,

Close #17469

I have added the percentile function in statistical functions of numpy frontend.

Please kindly check and review it.

Thank You

Dhruv Saxena

### CODE REVIEW

**Comment by ivy-leaves** [comment]
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

**Comment by dhruvsaxena11** [comment]
@ivy-leaves 

I have updated the description please kindly check.

Thank You

Dhruv Saxena

**Comment by dhruvsaxena11** [comment]
@ivy-leaves 

please kindly check my PR.

Thank You

**Review by dhruvsaxena11 — COMMENTED** [review]
@ivy-leaves 
Plz kindly review my PR as I updated the description. 

Thank you

**Comment by zhumakhan** [comment]
Hi, thanks for you PR!
Please also add test and decorators to the function like @to_ivy_arrays_and_back

**Comment by dhruvsaxena11** [comment]
@zhumakhan 

Hi,

I have updated the files and added the decorators to function and added the test.
please kindly check and suggest.

Thank You

**Review by dhruvsaxena11 — COMMENTED** [review]
@zhumakhan 

please kindly check and merge

**Comment by zhumakhan** [comment]
Could you please fix the implementation? Decorators in percentile.py file are not imported
Thanks!

**Comment by dhruvsaxena11** [comment]
@zhumakhan I updated the code please kindly check.

Thank You

**Comment by dhruvsaxena11** [comment]
@zhumakhan please kindly check and merge

File: ivy/functional/frontends/numpy/statistics/order_statistics.py [hunk 1]
```diff
@@ -35,6 +43,41 @@ def _cpercentile(N, percent, key=lambda x: x):
...
+        resultarray = ivy.percentile(nanlessarray, q)
+        return resultarray
+    elif axis == 1:
+        nanlessarrayofarrays = ivy.where(ivy.isnan(a), ivy.inf, a)
+        resultarray = ivy.percentile(nanlessarrayofarrays, q, axis=1, keepdims=keepdims)
+        return resultarray
+    elif axis == 0:
+        nanlessarrayofarrays = ivy.where(ivy.isnan(a), ivy.inf, a)
+        nanlessarrayofarrays = ivy.swapaxes(nanlessarrayofarrays, 0, 1)
+        resultarray = ivy.percentile(nanlessarrayofarrays, q, axis=0, keepdims=keepdims)
```
Comment (zhumakhan) [inline ivy/functional/frontends/numpy/statistics/order_statistics.py:78]: It seems `ivy.percentile` does not exist

File: ivy/functional/frontends/numpy/statistics/percentile.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    from_zero_dim_arrays_to_scalar,
+    handle_numpy_out,
+    
+)
+@handle_numpy_out
+@handle_numpy_dtype
+@to_ivy_arrays_and_back
+@from_zero_dim_arrays_to_scalar
+
+def percentile(
```
Comment (zhumakhan) [inline ivy/functional/frontends/numpy/statistics/percentile.py:16]: Defining `percentile`  in order_statistics.py file is enough. No need to duplicate.

**Review by zhumakhan — CHANGES_REQUESTED** [review]

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

File: ivy/functional/frontends/numpy/statistics/percentile.py (hunk 2 above)
Comment (dhruvsaxena11) [reply ivy/functional/frontends/numpy/statistics/percentile.py]: Done

**Comment by dhruvsaxena11** [comment]
Will working on it

File: ivy/functional/frontends/numpy/statistics/order_statistics.py (hunk 1 above)
Comment (dhruvsaxena11) [reply ivy/functional/frontends/numpy/statistics/order_statistics.py]: I have made some changes

File: ivy/functional/frontends/numpy/statistics/order_statistics.py [hunk 3]
```diff
@@ -1,5 +1,16 @@
...
+from ivy.functional.frontends.numpy.func_wrapper import (
+    to_ivy_arrays_and_back,
+    handle_numpy_dtype,
+    from_zero_dim_arrays_to_scalar,
+    handle_numpy_out,
+    
+)
+
 from ivy.functional.frontends.numpy.func_wrapper import (
     to_ivy_arrays_and_back,
```
Comment (zhumakhan) [inline ivy/functional/frontends/numpy/statistics/order_statistics.py:15]: Could you please remove this duplicate import?

File: ivy/functional/frontends/numpy/statistics/order_statistics.py [hunk 4]
```diff
@@ -37,6 +48,82 @@ def _cpercentile(N, percent, key=lambda x: x):
     d0 = key(N[int(f)]) * (c - k)
     d1 = key(N[int(c)]) * (k - f)
     return d0 + d1
+import ivy
```
Comment (zhumakhan) [inline ivy/functional/frontends/numpy/statistics/order_statistics.py:51]: duplicated import of ivy

File: ivy/functional/frontends/numpy/statistics/order_statistics.py [hunk 5]
```diff
@@ -1,5 +1,16 @@
 # global
 import ivy
+import numpy
```
Comment (zhumakhan) [inline ivy/functional/frontends/numpy/statistics/order_statistics.py:3]: please remove unused imports

File: ivy/functional/frontends/numpy/statistics/order_statistics.py [hunk 6]
```diff
@@ -1,5 +1,16 @@
 # global
 import ivy
+import numpy
+import ivy.numpy as np
```
Comment (zhumakhan) [inline ivy/functional/frontends/numpy/statistics/order_statistics.py:4]: This does not seem to exist

File: ivy/functional/frontends/numpy/statistics/order_statistics.py [hunk 7]
```diff
@@ -37,6 +48,82 @@ def _cpercentile(N, percent, key=lambda x: x):
     d0 = key(N[int(f)]) * (c - k)
     d1 = key(N[int(c)]) * (k - f)
     return d0 + d1
+import ivy
+
+
+@ivy.to_ivy_arrays_and_back
+@ivy.handle_numpy_out
```
Comment (zhumakhan) [inline ivy/functional/frontends/numpy/statistics/order_statistics.py:55]: `@handle_numpy_out` should work. `ivy` does not have `handle_numpy_out` attribute

**Review by zhumakhan — CHANGES_REQUESTED** [review]

**Comment by dhruvsaxena11** [comment]
@zhumakhan 

I have updated the code. Please kindly check and merge.

File: ivy/functional/frontends/numpy/statistics/order_statistics.py [hunk 8]
```diff
@@ -39,6 +43,82 @@ def _cpercentile(N, percent, key=lambda x: x):
...
+                arrayofpercentiles = []
+                for ii in nanlessarrayofarrays:
+                    arrayofpercentiles.append(_cpercentile(ii, i, interpolation))
+                resultarray.append(arrayofpercentiles)
+            return resultarray
+
+def _quantile_is_valid(q):
+    return ivy.all(ivy.logical_and(q >= 0, q <= 1))
+
+def _cpercentile(arr, q, interpolation='linear'):
```
Comment (zhumakhan) [inline ivy/functional/frontends/numpy/statistics/order_statistics.py:92]: It seems `_cpercentile` function already exists: https://github.com/unifyai/ivy/blob/main/ivy/functional/frontends/numpy/statistics/order_statistics.py#L13

**Review by zhumakhan — CHANGES_REQUESTED** [review]

**Comment by zhumakhan** [comment]
Hi, many tests are failing due to various reasons like shape related, unsupported dtypes and missing arguments to test function. Could you please thoroughly go through tests and fix them?
Thanks!

**Comment by ivy-seed** [comment]
This PR has been closed because it has been marked as stale for more than 7 days with no activity.

**Pull request closed without merging**